[MTIA][RNG] Add torch bindings for get and set default generator

diviramon opened the pull request:
Summary:
Implements MTIA RNG (Random Number Generator) integration into PyTorch following CUDA's architecture pattern. Adds default generators during device initialization and exposes them via `torch.mtia.default_generators`.

Key Changes

**C++ Infrastructure:**

*   Extended `MTIAHooksInterface` with `getDefaultGenerator()` and `getNewGenerator()` methods
*   Implemented hooks in `mtia_hooks.{h,cpp}` managing per-device RNG state
*   Added `createMTIAGenerator()` factory in `mtia_rng.{h,cpp}`
*   Modified `_mtia_init()` to create generators tuple mirroring CUDA implementation.

**Python Integration:**

*   Added `default_generators: tuple[torch._C.Generator] = ()` declaration to `torch/mtia/__init__.py`
*   Generators populated by C++ during lazy device initialization

**Architecture:**

*   Generators for same device share underlying RNG state. This is a key design choice for MTIA that users must be aware of.
*   Different devices maintain independent RNG states
*   Accessible via `torch.mtia.default_generators[device_idx]`

Test Plan:
buck2 test fbcode//mtia/host_runtime/torch_mtia/tests:test_mtia_rng_generators

https://www.internalfb.com/intern/testinfra/testrun/6192449762838044

Differential Revision: D88012916

patrick-toulme reviewed (approved):
Please get CI green and address one nit comment. Thanks!

andyanwang commented on torch/csrc/mtia/Module.cpp:
```diff
@@ -56,6 +58,30 @@ void initModule(PyObject* module) {
+        PyTuple_New(static_cast<Py_ssize_t>(num_devices));
+    for (const auto i : c10::irange(num_devices)) {
+      auto cast_gen = THPGenerator_initDefaultGenerator(
+          at::detail::getMTIAHooks().getDefaultGenerator(i));
```
@diviramon, sorry I was wrong. It turns out the failing OSS test test_cpp_extensions_mtia_backend uses the mocked MTIA hooks. And you probably need to add generator hooks to https://github.com/pytorch/pytorch/blob/main/test/cpp_extensions/mtia_extension.cpp#L139

andyanwang reviewed (approved):
lgtm

albanD reviewed (commented):
Sounds ok, only small nit

albanD reviewed (approved):
Thanks!

diviramon committed:
[MTIA][RNG] Add torch bindings for get and set default generator (#169271)

Summary:

Implements MTIA RNG (Random Number Generator) integration into PyTorch following CUDA's architecture pattern. Adds default generators during device initialization and exposes them via `torch.mtia.default_generators`.

Key Changes

**C++ Infrastructure:**

*   Extended `MTIAHooksInterface` with `getDefaultGenerator()` and `getNewGenerator()` methods
*   Implemented hooks in `mtia_hooks.{h,cpp}` managing per-device RNG state
*   Added `createMTIAGenerator()` factory in `mtia_rng.{h,cpp}`
*   Modified `_mtia_init()` to create generators tuple mirroring CUDA implementation.

**Python Integration:**

*   Added `default_generators: tuple[torch._C.Generator] = ()` declaration to `torch/mtia/__init__.py`
*   Generators populated by C++ during lazy device initialization

**Architecture:**

*   Generators for same device share underlying RNG state. This is a key design choice for MTIA that users must be aware of.
*   Different devices maintain independent RNG states
*   Accessible via `torch.mtia.default_generators[device_idx]`

Test Plan:
buck2 test fbcode//mtia/host_runtime/torch_mtia/tests:test_mtia_rng_generators

https://www.internalfb.com/intern/testinfra/testrun/6192449762838044

Reviewed By: klintqinami, andyanwang

Differential Revision: D88012916
